baseplate/lib/retry.py

```diff
@@ -124,14 +126,15 @@ def __init__(self, policy: RetryPolicy, base: float):
         self.subpolicy = policy
         self.base = base
 
-    def yield_attempts(self) -> Iterator[Optional[float]]:
-        for attempt, time_remaining in enumerate(self.subpolicy):
+    def yield_attempts(self) -> Iterator[Tuple["RetryPolicy", Optional[float]]]:
+        for attempt, (return_policy, value) in enumerate(self.subpolicy):
+            time_remaining = None
             if attempt > 0:
                 delay = self.base * 2.0 ** (attempt - 1.0)
-                if time_remaining:
-                    delay = min(delay, time_remaining)
-                    time_remaining -= delay
+                if value and isinstance(return_policy, TimeBudgetRetryPolicy):
```

Comment:
note: this fixes a bug in the current code.

in the current code, we always assume that `ExponentialBackoffRetryPolicy` is wrapping a `TimeBudgetRetryPolicy` by just assuming the iterator value is `time_remaining`. But that's actually not always true. Based on the implementation of `RetryPolicy.new`, if we set `backoff` but not `budget` it would be `ExponentialBackoffRetryPolicy` wrapping other subpolicies and the value from iterator is not really `time_remaining`.

cc @spladug